mbed_host_tests/host_tests_plugins/module_copy_pyocd.py

```diff
@@ -0,0 +1,139 @@
+from pyOCD.board import MbedBoard
+from intelhex import IntelHex
+import itertools
+
+
+def _ranges(i):
+    for a, b in itertools.groupby(enumerate(i), lambda (x, y): y - x):
+        b = list(b)
+        yield b[0][1], b[-1][1]
+
+
+class HostTestPluginCopyMethod_pyOCD(HostTestPluginBase):
+    # Plugin interface
+    name = 'HostTestPluginCopyMethod_pyOCD'
+    type = 'CopyMethod'
+    stable = True
+    capabilities = ['pyocd']
+    required_parameters = ['image_path', 'target_id']
+
+    def __init__(self):
+        """ ctor
+        """
+        HostTestPluginBase.__init__(self)
+
+    def setup(self, *args, **kwargs):
+        """ Configure plugin, this function should be called before plugin execute() method is used.
+        """
+        return True
+
+    def execute(self, capability, *args, **kwargs):
+        """! Executes capability by name
+
+        @param capability Capability name
+        @param args Additional arguments
+        @param kwargs Additional arguments
+        @return Capability call return value
+        """
+        if not self.check_parameters(capability, *args, **kwargs):
+            return False
+
+        if not kwargs['image_path']:
+            self.print_plugin_error("Error: image path not specified")
+            return False
+
+        if not kwargs['target_id']:
+            self.print_plugin_error("Error: Target ID")
+            return False
+
+        target_id = kwargs['target_id']
+        image_path = os.path.normpath(kwargs['image_path'])
+        with MbedBoard.chooseBoard(board_id=target_id) as board:
+            # Performance hack!
+            # Eventually pyOCD will know default clock speed
+            # per target
+            test_clock = 10000000
+            target_type = board.getTargetType()
+            if target_type == "nrf51":
+                # Override clock since 10MHz is too fast
+                test_clock = 1000000
+            if target_type == "ncs36510":
+                # Override clock since 10MHz is too fast
+                test_clock = 1000000
+
+            # Configure link
+            board.link.set_clock(test_clock)
+            board.link.set_deferred_transfer(True)
+
+            # Collect address, data pairs for programming
+            program_list = []
+            extension = os.path.splitext(image_path)[1]
+            if extension == '.bin':
+                # Binary file format
+                memory_map = board.target.getMemoryMap()
+                rom_region = memory_map.getBootMemory()
+                with open(image_path, "rb") as file_handle:
+                    program_data = file_handle.read()
+                program_list.append((rom_region.start, program_data))
+            elif extension == '.hex':
+                # Intel hex file format
+                ihex = IntelHex(image_path)
+                addresses = ihex.addresses()
+                addresses.sort()
+                data_list = list(_ranges(addresses))
```

Comment:
Perhaps some comment here saying "Get contiguous address ranges with start and end address".